system_metrics_collector/src/system_metrics_collector/linux_cpu_measurement_node.cpp

```diff
@@ -117,4 +118,36 @@ system_metrics_collector::ProcCpuData LinuxCpuMeasurementNode::makeSingleMeasure
   }
 }
 
+void LinuxCpuMeasurementNode::publishStatistics()
+{
+  if (publisher_ == nullptr) {
+    return;
+  }
+
+  const moving_average_statistics::StatisticData statistic_data = getStatisticsResults();
+  const double data[moving_average_statistics::STATISTICS_DATA_TYPES.size()] = {
+    statistic_data.average,
+    statistic_data.min,
+    statistic_data.max,
+    statistic_data.standard_deviation,
+    static_cast<double>(statistic_data.sample_count)
+  };
+
+  MetricsMessage msg = newMetricsMessage();
+  msg.metrics_source = "cpu_usage";
+  for (int i = 0; i < moving_average_statistics::STATISTICS_DATA_TYPES.size(); ++i) {
+    msg.statistics.emplace_back();
+    msg.statistics.back().data_type = moving_average_statistics::STATISTICS_DATA_TYPES[i];
+    msg.statistics.back().data = data[i];
+  }
+
+  publisher_->publish(msg);
+}
+
+bool LinuxCpuMeasurementNode::setupStart()
+{
+  last_measurement_ = ProcCpuData();
```

Comment:
Why do this in setup? Think of setup as a constructor helper, where no real work is done. Otherwise your measurement window is not accurate for the first sample.

Reply:
This is for the restart case. If you do `start()` -> `stop()` -> `start()`, you shouldn't have the first `MetricsMessage` you send after the second time you call `start()` use (stale) measurements from before the first time you called `stop()`.

Reply:
Moving this conversation to https://github.com/ros-tooling/system_metrics_collector/pull/16#discussion_r355054133.